caffe2/serialize/intermediate_model.h

```diff
@@ -0,0 +1,385 @@
+#pragma once
+
+#include <unordered_map>
+#include <stack>
+#include <string>
+
+#include <ATen/core/Allocator.h>
+#include <c10/util/Optional.h>
+
+#include "caffe2/core/common.h"
+#include "caffe2/serialize/inline_container.h"
+
+namespace caffe2 {
+  class TensorProto;
+  class NetDef;
+}
+
+namespace torch {
+  class ParameterDef;
+  class MethodDef;
+  class ModuleDef;
+  class ModelDef;
+}
+
+namespace at {
+namespace serialize {
+
+enum DeserializeMode {
+  // In EAGER mode, we load the file from the beginning, and eagarly load the content of tensors
+  EAGER = 1,
+  // In LAZY mode, we only load the last record of the file, which is the model metadata
+  // And the data of the tensor will be loaded later
+  LAZY = 2,
+};
+
+// multiple tensor may share the same content
+// SharedData contains:
+//    1) record id (i.e., the offset in the inline container)
+//    2) size, the size of the content
+//    3) data, in serialize, IntermediateModel does NOT own the data,
+//       in deserialize, the data pointer is returned by PyTorchFileReader,
+//       and IntermediateModel owns the data. The ownership later will be
+//       transferred to Tensor
+struct SharedData {
+  // constructor
+  explicit SharedData(uint64_t record_id) : recordId(record_id), dataPtr() {}
+  explicit SharedData(uint64_t record_id, at::DataPtr&& data_ptr, uint64_t size)
+    : recordId(record_id), dataPtr(std::move(data_ptr)), size(size) {}
+
+  c10::optional<uint64_t> recordId;
+  at::DataPtr dataPtr;
+  // For deserialize:
+  //   1) in EAGER mode, size information is avaiable immediately
+  //   2) in LAZY mode, until we load the tensor data, size won't be available
+  uint64_t size = 0;
```

Comment:
that's actually a minor nuisance for future mmap - we don't have the size of the shared storage without seeking to the record. We could store the storage size in the metadata protobuf as well. But I'm not sure it's worth doing now.

Reply:
Size information should be part of the metadata file, always. We don't want to have to jump around the file just to setup tensors. Furthermore, adding a single meta-data field to the records themselves will encourage more people to stash things with the records. Eventually this will devolve into having some fields next to the records and some in the meta-data file, which would be a very bad situation for versioning. My opinion is that all records should only be black boxes and that understanding them will require reading the meta-data file to find out anything else.

Reply:
Sure, I will add the size field to the metadata (i.e., protobuf) as well.